File: frontend/src/utils/response.ts
import {
  ApiError,
  ApiFail,
  ApiSuccess,
  ErrorCode,
} from "../types/response";

const now = () => new Date().toISOString();

export function success<T>(
  data: T,
  message = "요청이 성공적으로 처리되었습니다."
): ApiSuccess<T> {
  return {
    status: "SUCCESS",
    data,
    detail: { message, timestamp: now() },
  };
}

export function fail(code: ErrorCode | string, message: string): ApiFail {
  return {
    status: "FAIL",
    data: null,
    detail: { code, message, timestamp: now() },
  };
}

export function serverError(
  code: ErrorCode | string = "INTERNAL_SERVER_ERROR",
  message = "서버 내부 오류가 발생했습니다."
): ApiError {
  return {
    status: "ERROR",
    data: null,
    detail: { code, message, timestamp: now() },
  };
}

File: backend/repos/response.py
"""공통 API 응답 wrapper.

성공:
    { "status": "SUCCESS", "data": {...}, "detail": null, "timestamp": "..." }
실패:
    { "status": "FAIL", "data": null, "detail": {code, message}, "timestamp": "..." }

HTTP status와 body의 status는 분리한다.
body.status는 SUCCESS / FAIL 의 두 값만 사용한다.
"""

from datetime import datetime, timezone

from rest_framework import status as http_status
from rest_framework.response import Response


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def success(data=None, http: int = http_status.HTTP_200_OK) -> Response:
    return Response(
        {
            "status": "SUCCESS",
            "data": data if data is not None else {},
            "detail": None,
            "timestamp": _now_iso(),
        },
        status=http,
    )


def fail(code: str, message: str, http: int = http_status.HTTP_400_BAD_REQUEST) -> Response:
    return Response(
        {
            "status": "FAIL",
            "data": None,
            "detail": {"code": code, "message": message},
            "timestamp": _now_iso(),
        },
        status=http,
    )
